dev/TeachingTip/TeachingTip.cpp

```diff
@@ -0,0 +1,1540 @@
+                winrt::VisualStateManager::GoToState(*this, L"CloseButtonVisible"sv, false);
+            }
+        }
+        break;
+    }
+}
+
+void TeachingTip::UpdateDynamicBleedingContentPlacementToTop()
+{
+    if (BleedingImagePlacement() == winrt::TeachingTipBleedingImagePlacementMode::Auto)
+    {
+        winrt::VisualStateManager::GoToState(*this, L"BleedingContentTop"sv, false);
+        if (m_currentBleedingEffectivePlacementMode != winrt::TeachingTipBleedingImagePlacementMode::Top)
+        {
+            m_currentBleedingEffectivePlacementMode = winrt::TeachingTipBleedingImagePlacementMode::Top;
+            TeachingTipTestHooks::NotifyEffectiveBleedingPlacementChanged(*this);
+        }
+    }
+}
+
+void TeachingTip::UpdateDynamicBleedingContentPlacementToBottom()
+{
+    if (BleedingImagePlacement() == winrt::TeachingTipBleedingImagePlacementMode::Auto)
+    {
+        winrt::VisualStateManager::GoToState(*this, L"BleedingContentBottom"sv, false);
+        if (m_currentBleedingEffectivePlacementMode != winrt::TeachingTipBleedingImagePlacementMode::Bottom)
+        {
+            m_currentBleedingEffectivePlacementMode = winrt::TeachingTipBleedingImagePlacementMode::Bottom;
+            TeachingTipTestHooks::NotifyEffectiveBleedingPlacementChanged(*this);
+        }
+    }
+}
+
+void TeachingTip::OnIsOpenChanged()
+{
+    if (IsOpen())
+    {
+        m_lastCloseReason = winrt::TeachingTipCloseReason::Programmatic;
+        if (m_target)
+        {
+            m_currentTargetBounds = m_target.get().TransformToVisual(nullptr).TransformBounds({
+                0.0,
+                0.0,
+                static_cast<float>(m_target.get().as<winrt::FrameworkElement>().ActualWidth()),
+                static_cast<float>(m_target.get().as<winrt::FrameworkElement>().ActualHeight())
+                });
+        }
+        if (!m_popup)
+        {
+            m_popup.set(winrt::Popup());
+            m_popup.get().Child(*this);
+            m_popup.get().Closed({ this, &TeachingTip::OnPopupClosed });
+            if (IsLightDismissEnabled())
+            {
+                m_popup.get().IsLightDismissEnabled(true);
+            }
+        }
+        if (!m_contractAnimation)
+        {
+            CreateContractAnimation();
+        }        
+        if (!m_expandAnimation)
+        {
+            CreateExpandAnimation();
+        }
+
+        if (!m_popup.get().IsOpen())
+        {
+            //We are about to begin the process of trying to open the teaching tip, so notify that we are no longer idle.
+            if (m_isIdle)
+            {
+                m_isIdle = false;
+                TeachingTipTestHooks::NotifyIdleStatusChanged(*this);
+            }
+            m_popup.get().IsOpen(true);
+            if (m_rootGrid)
+            {
+                StartExpandAnimation();
+            }
+            else
+            {
+                m_startAnimationInOnApplyTemplate = true;
```

Comment:
Would it be better to just play animation on the next OnArrange call (or maybe at the end of Measure)? Then it would be the same time no matter what.